sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,479 @@
+    val validQuerySet = aType match {
+      case _: NumericType | DateType | TimestampType =>
+        val statsRange =
+          Range(aColStat.min, aColStat.max, aType).asInstanceOf[NumericRange]
+        hSet.map(e => numericLiteralToBigDecimal(e, aType, true)).
+          filter(e => e >= statsRange.min && e <= statsRange.max)
+
+      /** We assume the whole set since there is no min/max information for String/Binary type */
+      case StringType | BinaryType => hSet
+    }
+    if (validQuerySet.isEmpty) {
+      return 0.0
+    }
+
+    val newNdv = validQuerySet.size
+    val(newMax, newMin) = aType match {
+      case _: NumericType | DateType | TimestampType =>
+        val tmpSet: Set[Double] = validQuerySet.map(e => e.toString.toDouble)
+        (Some(tmpSet.max), Some(tmpSet.min))
+      case _ =>
+        (None, None)
+    }
+
+    if (update) {
+      val newStats = attrRef.dataType match {
+        case _: NumericType | DateType | TimestampType =>
+          aColStat.copy(distinctCount = newNdv, min = newMin,
+            max = newMax, nullCount = 0)
+        case StringType | BinaryType =>
+          aColStat.copy(distinctCount = newNdv, nullCount = 0)
+      }
+      mutableColStats += (attrRef.exprId -> newStats)
+    }
+
+    /**
+     * return the filter selectivity.  Without advanced statistics such as histograms,
+     * we have to assume uniform distribution.
+     */
+    math.min(1.0, validQuerySet.size / ndv.toDouble)
+  }
+
+  def evaluateBinaryForNumeric(
+      op: BinaryComparison,
+      attrRef: AttributeReference,
+      literal: Literal,
+      update: Boolean)
+    : Double = {
+
+    var percent = 1.0
+    val aColStat = mutableColStats(attrRef.exprId)
+    val ndv = aColStat.distinctCount
+    val statsRange =
+      Range(aColStat.min, aColStat.max, attrRef.dataType).asInstanceOf[NumericRange]
+
+    val literalValueBD = numericLiteralToBigDecimal(literal, attrRef.dataType)
+
+    /** determine the overlapping degree between predicate range and column's range */
+    val (noOverlap: Boolean, completeOverlap: Boolean) = op match {
+      case LessThan(l, r) =>
+        (literalValueBD <= statsRange.min, literalValueBD > statsRange.max)
+      case LessThanOrEqual(l, r) =>
+        (literalValueBD < statsRange.min, literalValueBD >= statsRange.max)
+      case GreaterThan(l, r) =>
+        (literalValueBD >= statsRange.max, literalValueBD < statsRange.min)
+      case GreaterThanOrEqual(l, r) =>
+        (literalValueBD > statsRange.max, literalValueBD <= statsRange.min)
+    }
+
+    if (noOverlap) {
+      percent = 0.0
+    } else if (completeOverlap) {
+      percent = 1.0
+    } else {
+      /** this is partial overlap case */
+      var newMax = aColStat.max
+      var newMin = aColStat.min
+      var newNdv = ndv
+      val literalToDouble = literalValueBD.toDouble
+      val maxToDouble = BigDecimal(statsRange.max).toDouble
+      val minToDouble = BigDecimal(statsRange.min).toDouble
+
+      /**

```

Comment:
please use `//` to document inline comments

`/** */` are reserved for classdocs/function docs.